deeposlandia/generator.py

```diff
@@ -5,54 +5,87 @@
 
 from keras.preprocessing.image import ImageDataGenerator
 
+def recover_label_id(img, label_config):
+    """Recover label ids starting from a typical RGB-colored image modeled as a
+    (width, height, 3)-shaped array
 
-def feature_detection_labelling(img, label_ids):
+    Parameters
+    ----------
+    img : np.array
+        Image data of shape (width, heigth, 3)
+    label_config : dict
+        Labels contained into the dataset
+
+    Returns
+    -------
+    np.array
+        Array of pixel labels of shape (width, heigth)
+    """
+    reshaped_img = img.reshape([-1, 3])
+    label_img = np.zeros(shape=reshaped_img.shape[0], dtype=reshaped_img.dtype)
+    for label in label_config:
+        label_img[np.all(label['color']==reshaped_img, axis=1)] = label['id']
+    return label_img.reshape(img.shape[:3])
+
+
+def feature_detection_labelling(img, label_config):
     """One-hot encoding for feature detection problem
 
     Parameters
     ----------
     img : numpy.array
         Batched input image data of size (batch_size, image_size, image_size, 1)
-    label_ids : list of integer
-        Number of classes contained into the dataset
+    label_config : dict
+        Labels contained into the dataset
+
     Returns
     -------
     numpy.array
         Label encoding, array of shape (batch_size, nb_labels)
     """
+    if not (img.shape[1] == img.shape[2] and
+            (len(img.shape) == 3 or len(img.shape) == 4)):
+        raise ValueError(("Wrong image shape. Please provide batched images "
+                          "with equal width and height dimensions."))
+    label_ids = [item['id'] for item in label_config if item['is_evaluate']]
     if not all(isinstance(item, (int, np.uint8, np.uint32, np.uint64)) for item in label_ids):
         raise ValueError(("List of label IDs must contains "
                           "integers: {}").format(label_ids))
+    if len(img.shape) == 4 and img.shape[3] > 1:
+        img = recover_label_id(img, label_config);
```

Comment:
The code blow up to pieces... I will add an `else` statement so as to raise an exception and catch the case in a cleaner way.
Here the "valid" method use is when we have batched RGB-images (format = [batch_size, im_size, im_size, nb_channel], with nb_channels == 3). We could debate on the acceptability of unbatched images, however a better way of doing is surely to prevent that, as we handle only batched images during dataset generator creations.

Reply:
I read the code again, and I guess we can run the method with only 3 dimensions, to the strict condition that we have [batch_size, im_size, im_size] (see first `if` statement in the method). Actually that's how Mapillary dataset is managed.

The following PR should harmonize the labelled image format (Mapillary labelled images will be stored as RGB instead of IDs), and such an `if` statement will become obsolete.